pkg/rbac/cacheValidation.go

```diff
@@ -0,0 +1,213 @@
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/klog/v2"
+)
+
+// Holds the objects needed to watch a kubernetes resource and trigger an action when a change is detected.
+type watchResource struct {
+	dynamicClient dynamic.Interface
+	gvr           schema.GroupVersionResource
+	onAdd         func(*unstructured.Unstructured)
+	onModify      func(*unstructured.Unstructured)
+	onDelete      func(*unstructured.Unstructured)
+}
+
+// Watch resources that could invalidate the cache.
+func (c *Cache) StartBackgroundValidation(ctx context.Context) {
+	klog.Info("Starting cache background validation.")
+
+	// Watch NAMESPACES
+	watchNamespaces := watchResource{
+		dynamicClient: c.shared.dynamicClient,
+		gvr:           schema.GroupVersionResource{Resource: "namespaces", Group: "", Version: "v1"},
+		onAdd:         c.namespaceAdded,
+		onModify:      nil, // Ignoring MODIFY because it doesn't affect the cached data.
+		onDelete:      c.namespaceDeleted,
+	}
+	go watchNamespaces.start(ctx)
+
+	// Watch ManagedClusters
+	watchManagedClusters := watchResource{
+		dynamicClient: c.shared.dynamicClient,
+		gvr: schema.GroupVersionResource{
+			Resource: "managedclusters",
+			Group:    "cluster.open-cluster-management.io",
+			Version:  "v1"},
+		onAdd:    c.managedClusterAdded,
+		onModify: nil, // Ignoring MODIFY because it doesn't affect the cached data.
+		onDelete: c.managedClusterDeleted,
+	}
+	go watchManagedClusters.start(ctx)
+
+	// Watch ManagedClusterAddon
+
+	// Watch ROLES
+
+	// Watch CLUSTERROLES
+
+	// Watch ROLEBINDINGS
+
+	// Watch CLUSTERRROLEBINDINGS
+
+	// Watch GROUPS
+
+	// Watch CRDS
+}
+
+// Start watching for changes to a resource and trigger the action to update the cache.
+func (w watchResource) start(ctx context.Context) {
+	for {
+		watch, watchError := w.dynamicClient.Resource(w.gvr).Watch(ctx, metav1.ListOptions{})
+		if watchError != nil {
+			klog.Warningf("Error watching %s, waiting 5 seconds before retry. Error: %s", w.gvr.String(), watchError)
+			time.Sleep(5 * time.Second) // Wait before retrying.
+			continue
+		}
+
+		defer watch.Stop()
+		klog.V(2).Infof("Watching resource: %s", w.gvr.String())
+
+		for {
+			select {
+			case <-ctx.Done():
+				klog.V(2).Info("Stopped watching resource. ", w.gvr.String())
+				watch.Stop()
+				return
+
+			case event := <-watch.ResultChan(): // Read events from the watch channel.
+				klog.V(6).Infof("Event: %s \tResource: %s  ", event.Type, w.gvr.String())
+				o, err := runtime.UnstructuredConverter.ToUnstructured(runtime.DefaultUnstructuredConverter, &event.Object)
+				if err != nil {
+					klog.Warningf("Error converting %s event.Object to unstructured.Unstructured. Error: %s",
+						w.gvr.Resource, err)
+				}
+				obj := &unstructured.Unstructured{Object: o}
```

Comment:
Can we convert to unstructured if error happens in the above line (if err!=nil)?

Reply:
If there's an error above, I assume that `o` will be an empty `map[string]struct{}` which could still be converted to unstructured, it will just be an empty object.

Reply:
So the error could potentially cause problems in the logic after, but II'm worried that returning could cause more harm.